media/js/base/translation-bar.es6.js

```diff
@@ -0,0 +1,110 @@
+        return false;
+    }
+
+    // Normalize language in the form of ab or ab-CD.
+    function normalize(lang) {
+        return lang.replace(/^(\w+)(?:-(\w+))?$/, function (m, p1, p2) {
+            return p1.toLowerCase() + (p2 ? '-' + p2.toUpperCase() : '');
+        });
+    }
+
+    // Returns boolean if lang matches target lang.
+    function isLangMatch(targetLang, lang) {
+        return (
+            targetLang === lang ||
+            targetLang === lang.split('-')[0] ||
+            targetLang.split('-')[0] === lang.split('-')[0]
+        );
+    }
+
+    // Return best matched available language, or `false` if none found.
+    function getBestMatch(acceptLangs, availableLangs) {
+        // Note: Do not use `forEach` here since you can't break the loop and
+        // will return the least best language.
+        for (const acceptLang of acceptLangs) {
+            for (const availableLang of availableLangs) {
+                if (isLangMatch(acceptLang, availableLang)) {
+                    return availableLang;
+                }
+            }
+        }
+        return false;
+    }
+
+    const pageLang = document.documentElement.lang;
+    const availableLangs = Array.from(
+        document.querySelectorAll('link[hreflang]'),
+        (link) => normalize(link.hreflang)
+    ).filter((e) => e !== 'x-DEFAULT');
+    const acceptLangs = navigator.languages.map((lang) => normalize(lang));
+    const bestLang = getBestMatch(acceptLangs, availableLangs);
+    const bestLink = document.querySelector(
+        `link[hreflang='${bestLang}']`
+    ).href;
+
+    let showBar = true;
+    if (
+        !bestLang ||
+        isLangMatch(pageLang, availableLangs[0]) ||
+        isLangMatch(pageLang, bestLang)
+    ) {
+        showBar = false;
+    }
+
+    function createBar(strings, link) {
+        /*
+         Build DOM of the translate bar.
+
+            <aside class="mzp-c-notification-bar">
+                <button class="mzp-c-notification-bar-button" type="button">Close notification</button>
+                <p><a href="{link}">View this page in your language</a></p>
+            </aside>
+
+        */
+
+        const fragment = new DocumentFragment();
+
+        const asideElem = document.createElement('aside');
+        asideElem.className = 'mzp-c-notification-bar';
+
+        const buttonElem = document.createElement('button');
+        buttonElem.className = 'mzp-c-notification-bar-button';
+        buttonElem.type = 'button';
+        buttonElem.appendChild(document.createTextNode(strings.close));
+
+        const pElem = document.createElement('p');
+        const aElem = document.createElement('a');
+        aElem.href = link;
+        aElem.appendChild(document.createTextNode(strings.message));
+        pElem.appendChild(aElem);
+
+        asideElem.appendChild(buttonElem);
+        asideElem.appendChild(pElem);
+
+        fragment.appendChild(asideElem);
+        barElem.appendChild(fragment);
```

Comment:
I could replace all of this by changing the JS request that gets the translated strings in the best language with an HTML fragment instead of a JSON doc with strings. Then the fragment could be appended to the `<div>`. Any considerations to think through here? I think the code would be simpler and easier to maintain if done in this way.

Reply:
That might work yeah. You could also use a JS string literal instead here and then insert that into the DOM in the same way you might with server rendered HTML. That might be simpler perhaps.

Reply:
I guess one other thing we might want to consider with this approach is avoiding any possible XSS vectors, by sanitising the translated strings to strip out any unexpected markup? Seems unlikely that would slip through in our L10n, but maybe worth thinking about.